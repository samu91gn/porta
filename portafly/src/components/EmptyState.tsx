import React from 'react'
import {
  Title,
  EmptyState,
  EmptyStateVariant,
  EmptyStateIcon,
  Bullseye,
  EmptyStateBody
} from '@patternfly/react-core'
import { CubesIcon, SearchIcon } from '@patternfly/react-icons'
import { IRow } from '@patternfly/react-table'

export interface ISimpleEmptyStateProps {
  msg: string
}

const SimpleEmptyState: React.FunctionComponent<ISimpleEmptyStateProps> = ({ msg }) => (
  <EmptyState variant={EmptyStateVariant.full}>
    <EmptyStateIcon icon={CubesIcon} />
    <Title headingLevel="h5" size="lg">
      {msg}
    </Title>
  </EmptyState>
)

type ICreateTableEmptyState = (title: string, body: string, button?: JSX.Element) => IRow[]

const CreateTableEmptyState: ICreateTableEmptyState = (title, body, button) => [{
  heightAuto: true,
  cells: [{
    props: { colSpan: 8 },
    title: (
      <Bullseye>
        <EmptyState variant={EmptyStateVariant.small}>
          <EmptyStateIcon icon={SearchIcon} />
          <Title headingLevel="h2" size="lg">
            {title}
          </Title>
          <EmptyStateBody>
            {body}
          </EmptyStateBody>
          {button}
        </EmptyState>
      </Bullseye>
    )
  }]
}]

export { SimpleEmptyState, CreateTableEmptyState }
