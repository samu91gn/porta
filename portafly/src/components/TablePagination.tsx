import React, { FunctionComponent, useMemo, useReducer } from 'react'
import { Pagination, OnSetPage, OnPerPageSelect } from '@patternfly/react-core'

interface IPagination {
  itemCount: number
  state: PaginationState
  dispatch: React.Dispatch<PaginationAction>
}

const perPageOptions = [
  { title: '5', value: 5 },
  { title: '20', value: 20 },
  { title: '50', value: 50 }
]

const TablePagination: FunctionComponent<IPagination> = ({
  itemCount,
  state,
  dispatch
}) => {
  const { perPage, page } = state

  const onSetPage: OnSetPage = (ev, newPage, _perPage, startIdx, endIdx) => {
    dispatch({
      type: 'setPage',
      payload: {
        page: newPage,
        startIdx,
        endIdx
      }
    })
  }

  const onPerPageSelect: OnPerPageSelect = (ev, newPerPage, newPage, startIdx, endIdx) => {
    dispatch({
      type: 'setPerPage',
      payload: {
        page: newPage,
        perPage: newPerPage,
        startIdx,
        endIdx
      }
    })
  }

  const isCompact = useMemo(() => (itemCount / perPage) < 3, [itemCount, perPage])

  return (
    <Pagination
      itemCount={itemCount}
      perPage={perPage}
      page={page}
      onSetPage={onSetPage}
      onPerPageSelect={onPerPageSelect}
      perPageOptions={perPageOptions}
      isCompact={isCompact}
    />
  )
}

type PaginationState = {
  page: number
  perPage: number
  startIdx?: number
  endIdx?: number
}

type PaginationAction = {
  type: 'setPage' | 'setPerPage'
  payload: Partial<PaginationState>
}

const paginationReducer = (state: PaginationState, action: PaginationAction):
PaginationState => {
  switch (action.type) {
    case 'setPage':
    case 'setPerPage':
      return { ...state, ...action.payload }
    default:
      return { ...state }
  }
}

const usePaginationReducer = () => useReducer(paginationReducer, {
  page: 0,
  perPage: perPageOptions[0].value,
  startIdx: 0,
  endIdx: perPageOptions[0].value
})

export { TablePagination, usePaginationReducer }
