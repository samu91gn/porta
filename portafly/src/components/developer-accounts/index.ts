export * from './ActionsDropdown'
export * from './BulkActionSelector'
export * from './SearchWidget'
export * from './DeveloperAccountsTable'
export * from './SendEmailModal'
export * from './ChangePlanModal'
export * from './ChangeStateModal'
