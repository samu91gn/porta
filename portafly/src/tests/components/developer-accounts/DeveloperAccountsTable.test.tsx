import React from 'react'

import { render } from 'tests/custom-render'
import { DeveloperAccountsTable } from 'components'
import { developerAccounts } from 'tests/examples'
import { RenderResult, fireEvent, within } from '@testing-library/react'

let wrapper: RenderResult

describe('when there are any accounts', () => {
  beforeEach(() => {
    wrapper = render(<DeveloperAccountsTable accounts={developerAccounts} />)
  })

  it('should render a table with accounts', () => {
    expect(wrapper.container.firstChild).toMatchSnapshot()
  })

  it('should be able to select one account', () => {
    const row = wrapper.getByText(developerAccounts[0].org_name).closest('tr') as HTMLElement
    const checkbox = within(row).getByRole('checkbox') as HTMLInputElement

    expect(checkbox.checked).toBe(false)
    fireEvent.click(checkbox)
    expect(checkbox.checked).toBe(true)
  })

  it('should be able to select all accounts', () => {
    const checkboxes = wrapper.getAllByRole('checkbox')
    const isChecked = (c: HTMLElement) => (c as HTMLInputElement).checked
    expect(checkboxes.filter(isChecked).length).toEqual(0)

    const checkbox = wrapper.getByTestId('developer-accounts-bulk-checkbox')
    fireEvent.click(checkbox)
    expect(checkboxes.filter(isChecked).length).toEqual(developerAccounts.length + 1)

    fireEvent.click(checkbox)
    expect(checkboxes.filter(isChecked).length).toEqual(0)
  })
})

describe('when there are NO accounts', () => {
  beforeEach(() => {
    wrapper = render(<DeveloperAccountsTable accounts={[]} />)
  })

  it('should render an empty view', () => {
    expect(wrapper.container.firstChild).toMatchSnapshot()
  })
})
