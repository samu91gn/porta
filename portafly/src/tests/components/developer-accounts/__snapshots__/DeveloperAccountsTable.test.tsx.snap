// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`when there are NO accounts should render an empty view 1`] = `
<div
  class="pf-c-empty-state"
>
  <svg
    aria-hidden="true"
    class="pf-c-empty-state__icon"
    fill="currentColor"
    height="1em"
    role="img"
    style="vertical-align: -0.125em;"
    viewBox="0 0 512 512"
    width="1em"
  >
    <path
      d="M488.6 250.2L392 214V105.5c0-15-9.3-28.4-23.4-33.7l-100-37.5c-8.1-3.1-17.1-3.1-25.3 0l-100 37.5c-14.1 5.3-23.4 18.7-23.4 33.7V214l-96.6 36.2C9.3 255.5 0 268.9 0 283.9V394c0 13.6 7.7 26.1 19.9 32.2l100 50c10.1 5.1 22.1 5.1 32.2 0l103.9-52 103.9 52c10.1 5.1 22.1 5.1 32.2 0l100-50c12.2-6.1 19.9-18.6 19.9-32.2V283.9c0-15-9.3-28.4-23.4-33.7zM358 214.8l-85 31.9v-68.2l85-37v73.3zM154 104.1l102-38.2 102 38.2v.6l-102 41.4-102-41.4v-.6zm84 291.1l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6zm240 112l-85 42.5v-79.1l85-38.8v75.4zm0-112l-102 41.4-102-41.4v-.6l102-38.2 102 38.2v.6z"
      transform=""
    />
  </svg>
  <h5
    class="pf-c-title pf-m-lg"
  >
    accounts_table.empty_state
  </h5>
</div>
`;

exports[`when there are any accounts should render a table with accounts 1`] = `
<div
  class="pf-c-data-toolbar"
  id="accounts-toolbar-top"
>
  <span
    hidden=""
    id="page-layout-table-column-management-action-toolbar-top-select-checkbox-label"
  >
    Choose one
  </span>
  <div
    class="pf-c-data-toolbar__content"
  >
    <div
      class="pf-c-data-toolbar__item"
    >
      <div
        class="pf-c-dropdown"
        id="checkbox_dropdown_id"
      >
        <div
          class="pf-c-dropdown__toggle pf-m-split-button"
        >
          <label
            class="pf-c-dropdown__toggle-check"
            for="developer-accounts-bulk-checkbox"
          >
            <input
              aria-invalid="false"
              aria-label="Select all"
              data-testid="developer-accounts-bulk-checkbox"
              id="developer-accounts-bulk-checkbox"
              type="checkbox"
            />
          </label>
          <button
            aria-expanded="false"
            aria-haspopup="true"
            aria-label="Select"
            class="pf-c-dropdown__toggle-button"
            id="pf-toggle-id-0"
            type="button"
          >
            
            <svg
              aria-hidden="true"
              class=""
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 320 512"
              width="1em"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                transform=""
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="pf-c-data-toolbar__item"
    >
      <div
        class="pf-c-dropdown"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          class="pf-c-dropdown__toggle pf-m-primary"
          id="pf-toggle-id-1"
          type="button"
        >
          <span
            class="pf-c-dropdown__toggle-text"
          >
            accounts_table.data_toolbar.bulk_actions.title
          </span>
          <svg
            aria-hidden="true"
            class="pf-c-dropdown__toggle-icon"
            fill="currentColor"
            height="1em"
            role="img"
            style="vertical-align: -0.125em;"
            viewBox="0 0 320 512"
            width="1em"
          >
            <path
              d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
              transform=""
            />
          </svg>
        </button>
      </div>
    </div>
    <div
      class="pf-c-data-toolbar__item"
    >
      <div
        class="pf-c-input-group"
      >
        <div
          class="pf-c-select"
        >
          <button
            aria-expanded="false"
            aria-haspopup="listbox"
            aria-labelledby="title-id pf-toggle-id-0"
            class="pf-c-select__toggle"
            id="pf-toggle-id-0"
            type="button"
          >
            <div
              class="pf-c-select__toggle-wrapper"
            >
              <span
                class="pf-c-select__toggle-icon"
              >
                <svg
                  aria-hidden="true"
                  fill="currentColor"
                  height="1em"
                  role="img"
                  style="vertical-align: -0.125em;"
                  viewBox="0 0 512 512"
                  width="1em"
                >
                  <path
                    d="M487.976 0H24.028C2.71 0-8.047 25.866 7.058 40.971L192 225.941V432c0 7.831 3.821 15.17 10.237 19.662l80 55.98C298.02 518.69 320 507.493 320 487.98V225.941l184.947-184.97C520.021 25.896 509.338 0 487.976 0z"
                    transform=""
                  />
                </svg>
              </span>
              <span
                class="pf-c-select__toggle-text"
              >
                accounts_table.col_group
              </span>
            </div>
            <svg
              aria-hidden="true"
              class="pf-c-select__toggle-arrow"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 320 512"
              width="1em"
            >
              <path
                d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                transform=""
              />
            </svg>
          </button>
        </div>
        <div
          class="pf-c-input-group"
        >
          <input
            aria-invalid="false"
            aria-label="accounts_table.data_toolbar.search_widget.text_input_aria_label"
            class="pf-c-form-control"
            placeholder="accounts_table.data_toolbar.search_widget.placeholder"
            type="search"
            value=""
          />
          <button
            aria-label="accounts_table.data_toolbar.search_widget.button_aria_label"
            class="pf-c-button pf-m-control"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 512 512"
              width="1em"
            >
              <path
                d="M505 442.7L405.3 343c-4.5-4.5-10.6-7-17-7H372c27.6-35.3 44-79.7 44-128C416 93.1 322.9 0 208 0S0 93.1 0 208s93.1 208 208 208c48.3 0 92.7-16.4 128-44v16.3c0 6.4 2.5 12.5 7 17l99.7 99.7c9.4 9.4 24.6 9.4 33.9 0l28.3-28.3c9.4-9.4 9.4-24.6.1-34zM208 336c-70.7 0-128-57.2-128-128 0-70.7 57.2-128 128-128 70.7 0 128 57.2 128 128 0 70.7-57.2 128-128 128z"
                transform=""
              />
            </svg>
          </button>
        </div>
      </div>
    </div>
    <div
      class="pf-c-data-toolbar__item pf-m-pagination pf-m-align-right-on-md"
    >
      <div
        class="pf-c-pagination"
        id="pagination-options-menu"
      >
        <div
          class="pf-c-pagination__total-items"
        >
          <b>
            1
             - 
            2
          </b>
           
          of 
          <b>
            2
          </b>
           
          
        </div>
        <div
          class="pf-c-options-menu"
        >
          <div
            class="pf-c-options-menu__toggle pf-m-plain pf-m-text"
          >
            <span
              class="pf-c-options-menu__toggle-text"
            >
              <b>
                1
                 - 
                2
              </b>
               
              of 
              <b>
                2
              </b>
               
              
            </span>
            <button
              aria-expanded="false"
              aria-label="Items per page"
              class="  pf-c-options-menu__toggle-button"
              id="pagination-options-menu-toggle"
              type="button"
            >
              <svg
                aria-hidden="true"
                class=""
                fill="currentColor"
                height="1em"
                role="img"
                style="vertical-align: -0.125em;"
                viewBox="0 0 320 512"
                width="1em"
              >
                <path
                  d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
                  transform=""
                />
              </svg>
            </button>
          </div>
        </div>
        <nav
          aria-label="Pagination"
          class="pf-c-pagination__nav"
        >
          <button
            aria-label="Go to first page"
            class="pf-c-button pf-m-plain"
            data-action="first"
            disabled=""
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 448 512"
              width="1em"
            >
              <path
                d="M223.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L319.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L393.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34zm-192 34l136 136c9.4 9.4 24.6 9.4 33.9 0l22.6-22.6c9.4-9.4 9.4-24.6 0-33.9L127.9 256l96.4-96.4c9.4-9.4 9.4-24.6 0-33.9L201.7 103c-9.4-9.4-24.6-9.4-33.9 0l-136 136c-9.5 9.4-9.5 24.6-.1 34z"
                transform=""
              />
            </svg>
          </button>
          <button
            aria-label="Go to previous page"
            class="pf-c-button pf-m-plain"
            data-action="previous"
            disabled=""
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 256 512"
              width="1em"
            >
              <path
                d="M31.7 239l136-136c9.4-9.4 24.6-9.4 33.9 0l22.6 22.6c9.4 9.4 9.4 24.6 0 33.9L127.9 256l96.4 96.4c9.4 9.4 9.4 24.6 0 33.9L201.7 409c-9.4 9.4-24.6 9.4-33.9 0l-136-136c-9.5-9.4-9.5-24.6-.1-34z"
                transform=""
              />
            </svg>
          </button>
          <div
            class="pf-c-pagination__nav-page-select"
          >
            <input
              aria-label="Current page"
              class="pf-c-form-control"
              disabled=""
              max="1"
              min="1"
              type="number"
              value="1"
            />
            <span
              aria-hidden="true"
            >
              of 
              1
            </span>
          </div>
          <button
            aria-label="Go to next page"
            class="pf-c-button pf-m-plain"
            data-action="next"
            disabled=""
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 256 512"
              width="1em"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34z"
                transform=""
              />
            </svg>
          </button>
          <button
            aria-label="Go to last page"
            class="pf-c-button pf-m-plain"
            data-action="last"
            disabled=""
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              role="img"
              style="vertical-align: -0.125em;"
              viewBox="0 0 448 512"
              width="1em"
            >
              <path
                d="M224.3 273l-136 136c-9.4 9.4-24.6 9.4-33.9 0l-22.6-22.6c-9.4-9.4-9.4-24.6 0-33.9l96.4-96.4-96.4-96.4c-9.4-9.4-9.4-24.6 0-33.9L54.3 103c9.4-9.4 24.6-9.4 33.9 0l136 136c9.5 9.4 9.5 24.6.1 34zm192-34l-136-136c-9.4-9.4-24.6-9.4-33.9 0l-22.6 22.6c-9.4 9.4-9.4 24.6 0 33.9l96.4 96.4-96.4 96.4c-9.4 9.4-9.4 24.6 0 33.9l22.6 22.6c9.4 9.4 24.6 9.4 33.9 0l136-136c9.4-9.2 9.4-24.4 0-33.8z"
                transform=""
              />
            </svg>
          </button>
        </nav>
      </div>
    </div>
    <div
      class="pf-c-data-toolbar__expandable-content"
      id="undefined-expandable-content-0"
    >
      <div
        class="pf-c-data-toolbar__group"
      />
    </div>
  </div>
  <div
    class="pf-c-data-toolbar__content pf-m-hidden"
    hidden=""
  >
    <div
      class="pf-c-data-toolbar__group"
    />
  </div>
</div>
`;
